arkouda/dataframe.py

```diff
@@ -977,6 +980,141 @@ def rename(self, mapper: Union[Callable, dict], inplace: bool = False) -> Union[
             return obj
         return None
 
+    @typechecked
+    def _rename_index(
+        self, mapper: Union[Callable, Dict], inplace: bool = False
+    ) -> Union[None, DataFrame]:
+        """
+        Rename indexes withing the dataframe
+
+        Parameters
+        ----------
+        mapper : callable or dict-like, Optional
+            Function or dictionary mapping existing indexes to new indexes.
+            Nonexistent names will not raise anerror.
+        inplace: bool
+            Default False. When True, perform the operation on the calling object.
+            When False, return a new object.
+        Returns
+        -------
+            DateFrame when `inplace=False`
+            None when `inplace=True`
+        See Also
+        -------
+            ak.DataFrame._rename_column
+            ak.DataFrame.rename
+        Notes
+        -----
+            This does not function exactly like pandas. The replacement value here must be
+            the same type as the existing value.
+        """
+        obj = self if inplace else self.copy()
+        # TODO - does the replacement need to be the same type????
+        if callable(mapper):
+            for i in range(obj.index.size):
+                oldval = obj.index[i]
+                newval = mapper(oldval)
+                if type(oldval) != type(newval):
+                    raise TypeError("Replacement value must have the same type as the original value")
+                obj.index.values[obj.index.values == oldval] = newval
```

Comment:
i thought of this a second too late for my first review, but would something like this work here? it might break the `Nonexistent names will not raise an error.` part, i haven't thought about it too hard
```python
if callable(mapper):
    obj.index.values = mapper(obj.index.values)
```

Reply:
I do not believe that we would be able to do something like this, because it relies on `mapper`  taking in a `pdarray` which is not necessarily a guarantee.

Reply:
oh good point. I wrote this right before i left for the day. All i really did to try it out was pull down the PR, make the change, and check that the tests still passed. I figured that was enough to leave a comment for it to be looked into more